feat: support to specify external pg instance as metadata storage

tisonkun opened the pull request:
This closes #1027.

Signed-off-by: tison <wander4096@gmail.com>

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -124,6 +127,7 @@ func init() {
 	rootCmd.PersistentFlags().BoolVar(&readonly, "readonly", false, "whether to run in read-only mode")
 	rootCmd.PersistentFlags().BoolVar(&demo, "demo", false, "whether to run using demo data")
 	rootCmd.PersistentFlags().BoolVar(&debug, "debug", false, "whether to enable debug level logging")
+	rootCmd.PersistentFlags().StringVar(&pgURL, "pg", "", "optional external postgresql instance url")
```
Let's add a comment for an example url string and also explain a bit background of the chosen url scheme.

tisonkun replied:
Do you mean a comment in code or the cmd output?

tianzhou replied:
Why not both :)

tianzhou replied:
BTW, one of the main reasons I'd like to see the format is because I can't think of a de facto URL format, so I am interested:)

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -327,13 +331,64 @@ func initBranding(ctx context.Context, settingService api.SettingService) error
-	ctx, cancel := context.WithCancel(ctx)
-	m.serverCancel = cancel
+func (m *Main) newDB() (*store.DB, error) {
+	if len(pgURL) > 0 {
```
Emit a log indicating whether we are using external or embed db

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -124,6 +127,7 @@ func init() {
 	rootCmd.PersistentFlags().BoolVar(&readonly, "readonly", false, "whether to run in read-only mode")
 	rootCmd.PersistentFlags().BoolVar(&demo, "demo", false, "whether to run using demo data")
 	rootCmd.PersistentFlags().BoolVar(&debug, "debug", false, "whether to enable debug level logging")
+	rootCmd.PersistentFlags().StringVar(&pgURL, "pg", "", "optional external postgresql instance url")
```
```suggestion
	rootCmd.PersistentFlags().StringVar(&pgURL, "pg", "", "optional external PostgreSQL instance url")
```

tianzhou reviewed (commented):
Thanks for taking a shot on this

d-bytebase commented:
I wonder if we could take hook up this place and skip embedded Postgres installation for an external source.

https://github.com/bytebase/bytebase/blob/main/bin/server/cmd/root.go#L341

tisonkun commented:
Thanks for your review. I'm addressing comments.

@d-bytebase yes. I can push a commit to do it. More user friendly changes can be made and I just omit them at the first place. Possibly you guys can review the whole config view exposed also.

tisonkun commented:
@d-bytebase @tianzhou comments addressed.

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -327,13 +341,69 @@ func initBranding(ctx context.Context, settingService api.SettingService) error
-func (m *Main) Run(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	m.serverCancel = cancel
+func isUseLocalDB() bool {
```
nit: useEmbeddedDB()

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -327,13 +341,69 @@ func initBranding(ctx context.Context, settingService api.SettingService) error
+
+func (m *Main) newDB() (*store.DB, error) {
+	if isUseLocalDB() {
+		return m.newLocalDB()
```
newEmbeddedDB()

tianzhou reviewed (changes requested):
The end state I am thinking of is to conditionally include the embedded PG binary. 

So we may introduce a build tag "omni" to indicate that we will build a self-contained binary (including embedded postgres binary and other future dependencies). Otherwise, we will just have a slim binary/docker image without those extra binaries.

And we may employ a similar strategy as our profile file to config the PostgreSQL endpoint.
https://sourcegraph.com/github.com/bytebase/bytebase/-/blob/bin/server/cmd/profile_dev.go

We can do them in separate PRs.

tianzhou commented:
> @tianzhou I'll fix inline comments tomorrow. For the conditional compile idea, I get your points and agree on them
> 
> Given that I cannot promise a continued period to work on this repo, I tend to do things in several small transactions instead of one large transaction :)
> 
> Adding this new feature won't affect a lot on conditional compile topic because we currently include those binaries and this is not a regression.
> 
> What do you think?

SGTM

tisonkun commented:
@tianzhou comments addressed. PTAL.

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -124,6 +129,7 @@ func init() {
 	rootCmd.PersistentFlags().BoolVar(&readonly, "readonly", false, "whether to run in read-only mode")
 	rootCmd.PersistentFlags().BoolVar(&demo, "demo", false, "whether to run using demo data")
 	rootCmd.PersistentFlags().BoolVar(&debug, "debug", false, "whether to enable debug level logging")
+	rootCmd.PersistentFlags().StringVar(&pgURL, "pg", "", "optional external PostgreSQL instance connection url; for example postgresql://user:secret@masterhost:5433/dbname?sslrootcert=cert")
```
Since the parameter is already pg, how about omitting the postgresql:// part?

tisonkun replied:
I'd prefer follow the existing standard (PostgreSQL already defines its connection URIs pattern) instead of make some unnecessary changes - it may looks convenient at first but actually a dialect to learn.

tianzhou replied:
Sounds reasonable.

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -273,23 +279,33 @@ func NewMain(activeProfile Profile, logger *zap.Logger) (*Main, error) {
-	if err != nil {
-		return nil, err
+	if useEmbeddedDB() {
+		logger.Info("Detecting and initializing local PostgreSQL instance...")
```
```suggestion
		logger.Info("Preparing embedded PostgreSQL instance...")
```

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -327,13 +343,69 @@ func initBranding(ctx context.Context, settingService api.SettingService) error
+}
+
+func (m *Main) newExternalDB() (*store.DB, error) {
+	m.l.Info("Initializing external PostgreSQL connection...", zap.String("pgURL", pgURL))
```
```suggestion
	m.l.Info("Establishing external PostgreSQL connection...", zap.String("pgURL", pgURL))
```

tianzhou replied:
If we want to print pgURL, we probably need to sanitize it as it contains the password

tisonkun replied:
I remove printing `pgURL` for now. It comes to my mind some brittle logics to do regexp matching and replace for an in-place sanitization, especially if there is syntax error. It isn't quite necessary to log `pgURL` here.

tisonkun replied:
Let me try to do it with `url.Redacted`, but it will only possible after parse succeeded.

tisonkun replied:
As 48232c2e5bd5a43eba4a9459a4999995562a9b80

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -327,13 +343,69 @@ func initBranding(ctx context.Context, settingService api.SettingService) error
+	return db, nil
+}
+
+func (m *Main) newLocalDB() (*store.DB, error) {
```
newEmbeddedDB

d-bytebase commented on bin/server/cmd/root.go:
```diff
@@ -91,6 +93,9 @@ var (
 	debug    bool
+	// pgURL must follow PostgreSQL connection URIs pattern.
+	// https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
+	pgURL string
```
I prefer that we take host/port/user/password information separately for a better user experience (less think). And not all parameters in pg urls are supported such as SSL that requires registration.

tianzhou reviewed (approved):
LGTM. Thanks!

@d-bytebase PTAL

tianzhou commented on bin/server/cmd/root.go:
```diff
@@ -91,6 +93,9 @@ var (
 	debug    bool
+	// pgURL must follow PostgreSQL connection URIs pattern.
+	// https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
+	pgURL string
```
FWIW, I did some research while reviewing this, that's why I was asking to provide the example as I didn't make up my mind before.

Many software does use separate flags. e.g.

GitLab does separate them
https://docs.gitlab.com/omnibus/settings/database.html#configure-postgresql-block

Sourcegraph also separates them
https://docs.sourcegraph.com/admin/external_services/postgres

So is temporal
https://docs.temporal.io/docs/server/configuration/#persistence
https://docs.temporal.io/docs/server/versions-and-dependencies/#upgrading-mysql--postgresql-schema

While I am also in the Go DSN / JDBC camp to put all into a single connection string like how this PR is implemented.

d-bytebase replied:
My personal preference is to use separate flags instead of taking a whole DSN string for a better user experience. Users don't need to look up the format of Postgres DSN.

tisonkun replied:
If we configure all this options in a config file, said a yaml or toml file, I'll prefer to use separate flags, e.g.,

```toml
[metadata.external]
enable = true
type = "postgresql"
user = "..."
passwd = "..."
# ...
```

However, it's in the command line and the parts are grouped to be present or absent (whether to provide an external pgURL is optional). Thus, if we flatten all flags to the command line it'll be quite clumsy and the composition is brittle.

tianzhou replied:
The downside of having separate flags is the user needs to learn to feed the values to the appropriate flags respectively.

Our situation is worse, e.g. we already have port, frontend-port, then we will have pg-port.

tisonkun replied:
@d-bytebase is it possible for `cobra` to group flags? If it's possible, then I can try to make an alternative for preview.

tianzhou replied:
TBH, though it's a bit struggle, I prefer the current way to have a single flag.

d-bytebase replied:
Let's take a try for now.

tisonkun commented on bin/server/cmd/root.go:
```diff
@@ -273,23 +279,33 @@ func NewMain(activeProfile Profile, logger *zap.Logger) (*Main, error) {
-	pgInstance, err := postgres.Install(resourceDir, pgDataDir, activeProfile.pgUser)
-	if err != nil {
-		return nil, err
+	if useEmbeddedDB() {
```
resolved at 3b6421bfc5a94e106d0980948a3fd08e88f4d597

tianzhou reviewed (approved):
LGTM

tianzhou commented:
Thanks for the contribution ^_^

d-bytebase commented:
Thanks for the contribution! We will take some testing with external databases such as RDS aws.
